README
USB CCID IFD Handler
====================

This package provides the source code for a generic USB CCID (Chip/Smart
Card Interface Devices) and ICCD (Integrated Circuit(s) Card Devices)
driver. See the USB CCID [1] and ICCD [2] specifications from the USB
working group.

[1] http://www.usb.org/developers/devclass_docs/DWG_Smart-Card_CCID_Rev110.pdf
[2] http://www.usb.org/developers/devclass_docs/DWG_Smart-Card_USB-ICC_ICCD_rev10.pdf

Authors:
========

- Ludovic Rousseau <ludovic.rousseau@free.fr>
- Carlos Prados for the PPS and ATR parsing code (taken from his
  towitoto driver) in towitoko/ directory.
- Olaf Kirch for the T=1 TPDU code (from the OpenCT package) in openct/
  directory. I (Ludovic Rousseau) greatly improved this code.


CCID and ICCD readers:
======================

A reader can be in one of these list:
- supported
    See http://pcsclite.alioth.debian.org/ccid/supported.html
- should working
    See http://pcsclite.alioth.debian.org/ccid/shouldwork.html
- unsupported
    See http://pcsclite.alioth.debian.org/ccid/unsupported.html
- disabled
    See http://pcsclite.alioth.debian.org/ccid/disabled.html


Supported operating systems:
============================

- GNU/Linux (libusb 1.0)
- MacOS X/Darwin (libusb 1.0)

See also http://pcsclite.alioth.debian.org/ccid.html for more
information.


Debug informations:
===================

The driver uses the debug function provided by pcscd. So if pcscd sends
its debug to stdout (pcscd --foreground) then the CCID driver will also
send its debug to stdout. If pcscd sends its debug to syslog (by
default) then the CCID driver will also send its debug to syslog.

You can change the debug level using the Info.plist configuration file.
The Info.plist is installed, by default, in
/usr/local/pcsc/drivers/ifd-ccid.bundle/Contents/Info.plist
or set the environment variable LIBCCID_ifdLogLevel.

The debug level is set in the ifdLogLevel field. It is a binary OR
combinaison of 4 different levels.
- 1: critical: important error messages
- 2: info:     informative messages like what reader was detected
- 4: comm:     a dump of all the bytes exchanged between the host and the
               reader
- 8: periodic: periodic info when pcscd test if a card is present (every
               1/10 of a second)

By default the debug level is set to 3 (1 + 2) and correspond to the
critical and info levels.

You have to restart the driver so it read the configuration file again
and use the new debug level value.  To restart the driver you just need
to unplug all your CCID readers so the the driver is unloaded and then
replug your readers. You can also restart pcscd.


Voltage selection
=================

You can change the voltage level using the Info.plist configuration
file.  The Info.plist is installed, by default, in
/usr/local/pcsc/drivers/ifd-ccid.bundle/Contents/Info.plist

The voltage level is set in the ifdDriverOptions field. It is a binary OR
combinaison of 4 different levels.

-  0: power on the card at 5V (default value)
- 16: power on the card at 3V and, if 3V fails then use 5V
- 32: power on the card at 1.8V, then 3V and then 5V
- 48: let the reader decide

By default the voltage level is set to 0 and correspond to 5V.

You have to restart the driver so it read the configuration file again
and use the new debug level value.  To restart the driver you just need
to unplug all your CCID readers so the the driver is unloaded and then
replug your readers.  You can also restart pcscd.


Licence:
========

  This library is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or (at
your option) any later version.

  This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser
General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
along with this library; if not, write to the Free Software Foundation,
Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


History:
========

1.4.16 - 23 March 2014, Ludovic Rousseau
    - Add support of
      . Crypto Stick Crypto Stick v1.4
      . Hewlett Packard USB Smartcard CCID Keyboard
      . IID AT90S064 CCID READER
      . INSIDE Secure VaultIC 405 Smart Object
      . INSIDE Secure VaultIC 441 Smart Object
      . Microchip SEC1110
      . Microchip SEC1210
      . Watchdata W5181
    - Add support of DRIVER_OPTION_DISABLE_PIN_RETRIES
      The Gemalto pinpad reader sends a VERIFY command with no PIN value
      in order to retreive the remaining retries from the card.  Some
      cards (like the OpenPGP card) do not support this.
      It is now possible to disable this behavior from the Gemalto
      Pinpad and Covadis Véga Alpha.
    - add support of WTX received before SW during Secure Pin Entry Verify
      The Swiss health care card sends a WTX request before returning
      the SW code. If the reader is in TPDU and the card is in T=1 the
      driver must manage the request itself.


1.4.15 - 14 February 2014, Ludovic Rousseau
    - Add support of
      . DUALi DRAGON NFC READER
      . Feitian bR301
      . Gemalto CR30 reader in serial communication
      . Gemalto Ezio Shield Pro SC
      . IIT E.Key Almaz-1C
    - PIN_MODIFY_STRUCTURE & PIN_VERIFY_STRUCTURE: Fix calculation of
      the command length after pcsc-lite 1.8.9 (October 2013) changed
      the PCSC/reader.h header
    - Add specific PIN min (0) & max (25) sizes for SmartTerminal
      ST-2xxx
    - Do not get the data rates if bNumDataRatesSupported = 0
    - Support Gemalto features for pinpad readers MinimumPINSize,
      MaximumPINSize and bEntryValidationCondition are fetched from the
      reader firmware
    - disable (broken) pinpad for Fujitsu SmartCase KB SCR eSIG
    - examples/scardcontrol.c:
      . Parse codes returned by a pinpad (as SW1/SW2)
        Known codes for now are:
        0x9000: Success
        0x6400: Timeout
        0x6401: Cancelled by user
        0x6402: PIN mismatch
        0x6403: Too short or too long PIN
      . Retrieve min and max PIN sizes from the driver
      . Retrieve bEntryValidationCondition from the driver
    - be more strict for bInterfaceClass = 255 by also checking extra_length
    - some minor bugs removed


1.4.14 - 25 November 2013, Ludovic Rousseau
    - Add support of
      . Gemalto GemCore SIM Pro firmware 2.0 (using USB)
    - report FEATURE_IFD_PIN_PROPERTIES only for pinpad readers
    - Generalize the management of (old) readers with bDeviceClass = 0xFF
    - some minor bugs removed


1.4.13 - 9 October 2013, Ludovic Rousseau
    - Add support of
      . Access IS ePassport Reader
      . Planeta RC700-NFC CCID
    - Add support of Windows value for CM_IOCTL_GET_FEATURE_REQUEST
      Windows uses 0x313520 for SCARD_CTL_CODE(3400) pcsc-lite uses
      0x42000D48 for SCARD_CTL_CODE(3400)
      RDP aplications (like rdesktop) will convert SCardControl()
      commands from a Windows application (so using 0x313520) to
      pcsc-lite.
    - fix multi-slot support for card movement notification (introduced
      in 1.4.12)
    - Mac OS X: differentiate each libccid library by the dynamic linker
      using --prefix=/fake/$BUNDLE_ID
    - some minor bugs removed


1.4.12 - 12 August 2013, Ludovic Rousseau
    - Add support of
      . HID OMNIKEY 5127 CK
      . HID OMNIKEY 5326 DFR
      . HID OMNIKEY 5427 CK
      . Ingenico WITEO USB Smart Card Reader (Base and Badge)
      . SecuTech SecuTech Token
    - Add support of card movement notifications for multi-slot readers
    - Check libusb is at least at version 1.0.8
    - Get the serialconfdir value from pcsc-lite pkg config instead of
      using $(DESTDIR)/$(sysconfdir)/reader.conf.d/
    - Disable class driver on Mac OS X
    - Update the bundle name template to include the vendor name
    - some minor bugs removed


1.4.11 - 12 June 2013, Ludovic Rousseau
    - Add support of
       . Gemalto IDBridge CT30
       . Gemalto IDBridge K30
       . SCM Microsystems Inc. SCL010 Contactless Reader
       . SCM Microsystems Inc. SDI011 Contactless Reader
       . THRC reader
    - Better management of time extension requests
    - parse: better support of devices with bInterfaceClass = 0xFF
    - udev rule file: Remove setting group to pcscd, remove support of
      Linux kernel < 2.6.35 for auto power up management
    - some minor bugs removed


1.4.10 - 16 April 2013, Ludovic Rousseau
    - Add support of
       . ACS APG8201 USB Reader with PID 0x8202
       . GIS Ltd SmartMouse USB
       . Gemalto IDBridge K3000
       . Identive CLOUD 2700 F Smart Card Reader
       . Identive CLOUD 2700 R Smart Card Reader
       . Identive CLOUD 4500 F Dual Interface Reader
       . Identive CLOUD 4510 F Contactless + SAM Reader
       . Identive CLOUD 4700 F Dual Interface Reader
       . Identive CLOUD 4710 F Contactless + SAM Reader
       . Inside Secure AT90SCR050
       . Inside Secure AT90SCR100
       . Inside Secure AT90SCR200
       . SCR3310-NTTCom USB SmartCard Reader
       . SafeTech SafeTouch
       . SpringCard H512 Series
       . SpringCard H663 Series
       . SpringCard NFC'Roll

ChangeLog
2014-03-23  Ludovic Rousseau

	* [r6880] README, configure.ac: Release 1.4.16
	* [r6879] src/commands.c: Fix compiler warning
	  
	  commands.c: In function `SecurePINVerify': commands.c:529:9:
	  warning: variable `n' set but not used
	  [-Wunused-but-set-variable] int n, oldReadTimeout; ^
	* [r6878] src/commands.c: Fix compiler warning
	  
	  commands.c:511:19: warning: | has lower precedence than ==; ==
	  will be evaluated first [-Wparentheses] if (T1_S_BLOCK | T1_S_WTX
	  == RxBuffer[PCB]) ^~~~~~~~~~~~~~~~~~~~~~~~~~~ commands.c:511:19:
	  note: place parentheses around the '==' expression to silence
	  this warning if (T1_S_BLOCK | T1_S_WTX == RxBuffer[PCB]) ^
	  ~~~~~~~~~~~~~~~~~~~~~~~~~ commands.c:511:19: note: place
	  parentheses around the | expression to evaluate it first if
	  (T1_S_BLOCK | T1_S_WTX == RxBuffer[PCB]) ~~~~~~~~~~~^~~~~~~~~~ 1
	  warning generated.
	* [r6877] src/commands.c: SecurePINVerify: add support of WTX
	  received before SW
	  
	  The Swiss health care card sends a WTX request before returning
	  the SW code. If the reader is in TPDU and the card is in T=1 the
	  driver must manage the request itself.
	* [r6876] src/ccid_ifdhandler.h: Add
	  DRIVER_OPTION_DISABLE_PIN_RETRIES definition
	  
	  This is a missing for revision 6875

2014-03-22  Ludovic Rousseau

	* [r6875] src/Info.plist.src, src/ccid.c: Add support of
	  DRIVER_OPTION_DISABLE_PIN_RETRIES
	  
	  The Gemalto pinpad reader sends a VERIFY command with no PIN
	  value in order to retreive the remaining retries from the card.
	  Some cards (like the OpenPGP card) do not support this.
	  
	  It is now possible to disable this behavior from the Gemalto
	  Pinpad and Covadis Véga Alpha.
	  
	  Closes Alioth bug
	  https://alioth.debian.org/tracker/index.php?func=detail&aid=314005&group_id=30105&atid=410088
	* [r6874] src/ccid_usb.c: InterruptRead(): log the libusb errors
	  
	  Allow an easier debug

2014-03-11  Ludovic Rousseau

	* [r6873] readers/IID_AT90S064_CCID_READER.txt,
	  readers/supported_readers.txt: Add IID AT90S064 CCID READER

2014-03-09  Ludovic Rousseau

	* [r6871] readers/Watchdata_W5181.txt,
	  readers/supported_readers.txt: Add Watchdata W5181

2014-03-08  Ludovic Rousseau

	* [r6869] src/openct/proto-t1.c: t1_transceive(): add a new test on
	  retries
	  
	  The driver should not send R-blocks infinitely.
	  
	  The problem was detected by a T=1 card on a TPDU pinpad reader
	  with a VERIFY command that send a WTX request before returning
	  the SW code.
	* [r6868] src/openct/proto-t1.c: t1_transceive(): check retries is
	  is positive
	  
	  The test was "retries == 0" but it may happen that retries
	  becomes negative. Then the counter will be decreasing
	  "indefinitely".
	  
	  The new test is "retries <= 0" to also get the negative values
	  and resync if needed.
	  
	  The problem was detected by a T=1 card on a TPDU pinpad reader
	  with a VERIFY command that send a WTX request before returning
	  the SW code.

2014-03-07  Ludovic Rousseau

	* [r6866] readers/INSIDE_Secure_VaultIC_441_Smart_Object.txt,
	  readers/supported_readers.txt: Add INSIDE Secure VaultIC 441
	  Smart Object
	* [r6865] readers/INSIDE_Secure_VaultIC_405_Smart_Object.txt,
	  readers/supported_readers.txt: Add INSIDE Secure VaultIC 405
	  Smart Object
	* [r6864] readers/supported_readers.txt: Replace HP by Hewlett
	  Packard

2014-02-26  Ludovic Rousseau

	* [r6862] NEWS: Indicates the changes are in README.

2014-02-23  Ludovic Rousseau

	* [r6860] readers/Chicony_HP_USB_Smartcard_CCID_Keyboard.txt,
	  readers/supported_readers.txt: Add HP USB Smartcard CCID Keyboard

2014-02-18  Ludovic Rousseau

	* [r6857] readers/Microchip_SEC1210.txt,
	  readers/supported_readers.txt: Add Microchip SEC1210
	* [r6856] readers/Microchip_SEC1110.txt,
	  readers/supported_readers.txt: Add Microchip SEC1110
	* [r6855] readers/Crypto_Stick_Crypto_Stick_v1.4.txt,
	  readers/supported_readers.txt: Add Crypto Stick Crypto Stick v1.4

2014-02-14  Ludovic Rousseau

	* [r6846] README, configure.ac: Release 1.4.15
	* [r6845] src/ccid.c, src/ccid.h: Disable pinpad for Fujitsu
	  SmartCase KB SCR eSIG
	  
	  Thanks to Maarten for the bug report and patch.

2014-02-08  Ludovic Rousseau

	* [r6839] readers/Feitian_bR301.txt, readers/supported_readers.txt:
	  Add Feitian bR301

2014-02-03  Ludovic Rousseau

	* [r6835] readers/Gemalto_Ezio_Shield_Pro_SC.txt,
	  readers/supported_readers.txt: Add Gemalto Ezio Shield Pro SC

2014-01-28  Ludovic Rousseau

	* [r6833] src/reader.conf.in: Remove the optional CHANNELID
	  
	  Use DEVICENAME only instead.
	* [r6832] src/ccid.h, src/ccid_serial.c, src/reader.conf.in: Add

MacOSX/reader.h
/*
 * MUSCLE SmartCard Development ( http://www.linuxnet.com )
 *
 * Copyright (C) 1999-2005
 *  David Corcoran <corcoran@linuxnet.com>
 * Copyright (C) 2005-2009
 *  Ludovic Rousseau <ludovic.rousseau@free.fr>
 *
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. The name of the author may not be used to endorse or promote products
   derived from this software without specific prior written permission.

Changes to this license can be made only by the copyright author with
explicit written consent.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id: reader.h 6822 2014-01-13 15:21:25Z rousseau $
 */

/**
 * @file
 * @brief This keeps a list of defines shared between the driver and the application
 */

#ifndef __reader_h__
#define __reader_h__

/*
 * Tags for requesting card and reader attributes
 */

#define SCARD_ATTR_VALUE(Class, Tag) ((((ULONG)(Class)) << 16) | ((ULONG)(Tag)))

#define SCARD_CLASS_VENDOR_INFO     1   /**< Vendor information definitions */
#define SCARD_CLASS_COMMUNICATIONS  2   /**< Communication definitions */
#define SCARD_CLASS_PROTOCOL        3   /**< Protocol definitions */
#define SCARD_CLASS_POWER_MGMT      4   /**< Power Management definitions */
#define SCARD_CLASS_SECURITY        5   /**< Security Assurance definitions */
#define SCARD_CLASS_MECHANICAL      6   /**< Mechanical characteristic definitions */
#define SCARD_CLASS_VENDOR_DEFINED  7   /**< Vendor specific definitions */
#define SCARD_CLASS_IFD_PROTOCOL    8   /**< Interface Device Protocol options */
#define SCARD_CLASS_ICC_STATE       9   /**< ICC State specific definitions */
#define SCARD_CLASS_SYSTEM     0x7fff   /**< System-specific definitions */

#define SCARD_ATTR_VENDOR_NAME SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_INFO, 0x0100) /**< Vendor name. */
#define SCARD_ATTR_VENDOR_IFD_TYPE SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_INFO, 0x0101) /**< Vendor-supplied interface device type (model designation of reader). */
#define SCARD_ATTR_VENDOR_IFD_VERSION SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_INFO, 0x0102) /**< Vendor-supplied interface device version (DWORD in the form 0xMMmmbbbb where MM = major version, mm = minor version, and bbbb = build number). */
#define SCARD_ATTR_VENDOR_IFD_SERIAL_NO SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_INFO, 0x0103) /**< Vendor-supplied interface device serial number. */
#define SCARD_ATTR_CHANNEL_ID SCARD_ATTR_VALUE(SCARD_CLASS_COMMUNICATIONS, 0x0110) /**< DWORD encoded as 0xDDDDCCCC, where DDDD = data channel type and CCCC = channel number */
#define SCARD_ATTR_ASYNC_PROTOCOL_TYPES SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0120) /**< FIXME */
#define SCARD_ATTR_DEFAULT_CLK SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0121) /**< Default clock rate, in kHz. */
#define SCARD_ATTR_MAX_CLK SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0122) /**< Maximum clock rate, in kHz. */
#define SCARD_ATTR_DEFAULT_DATA_RATE SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0123) /**< Default data rate, in bps. */
#define SCARD_ATTR_MAX_DATA_RATE SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0124) /**< Maximum data rate, in bps. */
#define SCARD_ATTR_MAX_IFSD SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0125) /**< Maximum bytes for information file size device. */
#define SCARD_ATTR_SYNC_PROTOCOL_TYPES SCARD_ATTR_VALUE(SCARD_CLASS_PROTOCOL, 0x0126) /**< FIXME */
#define SCARD_ATTR_POWER_MGMT_SUPPORT SCARD_ATTR_VALUE(SCARD_CLASS_POWER_MGMT, 0x0131) /**< Zero if device does not support power down while smart card is inserted. Nonzero otherwise. */
#define SCARD_ATTR_USER_TO_CARD_AUTH_DEVICE SCARD_ATTR_VALUE(SCARD_CLASS_SECURITY, 0x0140) /**< FIXME */
#define SCARD_ATTR_USER_AUTH_INPUT_DEVICE SCARD_ATTR_VALUE(SCARD_CLASS_SECURITY, 0x0142) /**< FIXME */
#define SCARD_ATTR_CHARACTERISTICS SCARD_ATTR_VALUE(SCARD_CLASS_MECHANICAL, 0x0150) /**< DWORD indicating which mechanical characteristics are supported. If zero, no special characteristics are supported. Note that multiple bits can be set */

#define SCARD_ATTR_CURRENT_PROTOCOL_TYPE SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0201) /**< FIXME */
#define SCARD_ATTR_CURRENT_CLK SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0202) /**< Current clock rate, in kHz. */
#define SCARD_ATTR_CURRENT_F SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0203) /**< Clock conversion factor. */
#define SCARD_ATTR_CURRENT_D SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0204) /**< Bit rate conversion factor. */
#define SCARD_ATTR_CURRENT_N SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0205) /**< Current guard time. */
#define SCARD_ATTR_CURRENT_W SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0206) /**< Current work waiting time. */
#define SCARD_ATTR_CURRENT_IFSC SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0207) /**< Current byte size for information field size card. */
#define SCARD_ATTR_CURRENT_IFSD SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0208) /**< Current byte size for information field size device. */
#define SCARD_ATTR_CURRENT_BWT SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x0209) /**< Current block waiting time. */
#define SCARD_ATTR_CURRENT_CWT SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x020a) /**< Current character waiting time. */
#define SCARD_ATTR_CURRENT_EBC_ENCODING SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x020b) /**< Current error block control encoding. */
#define SCARD_ATTR_EXTENDED_BWT SCARD_ATTR_VALUE(SCARD_CLASS_IFD_PROTOCOL, 0x020c) /**< FIXME */

#define SCARD_ATTR_ICC_PRESENCE SCARD_ATTR_VALUE(SCARD_CLASS_ICC_STATE, 0x0300) /**< Single byte indicating smart card presence */
#define SCARD_ATTR_ICC_INTERFACE_STATUS SCARD_ATTR_VALUE(SCARD_CLASS_ICC_STATE, 0x0301) /**< Single byte. Zero if smart card electrical contact is not active; nonzero if contact is active. */
#define SCARD_ATTR_CURRENT_IO_STATE SCARD_ATTR_VALUE(SCARD_CLASS_ICC_STATE, 0x0302) /**< FIXME */
#define SCARD_ATTR_ATR_STRING SCARD_ATTR_VALUE(SCARD_CLASS_ICC_STATE, 0x0303) /**< Answer to reset (ATR) string. */
#define SCARD_ATTR_ICC_TYPE_PER_ATR SCARD_ATTR_VALUE(SCARD_CLASS_ICC_STATE, 0x0304) /**< Single byte indicating smart card type */

#define SCARD_ATTR_ESC_RESET SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_DEFINED, 0xA000) /**< FIXME */
#define SCARD_ATTR_ESC_CANCEL SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_DEFINED, 0xA003) /**< FIXME */
#define SCARD_ATTR_ESC_AUTHREQUEST SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_DEFINED, 0xA005) /**< FIXME */
#define SCARD_ATTR_MAXINPUT SCARD_ATTR_VALUE(SCARD_CLASS_VENDOR_DEFINED, 0xA007) /**< FIXME */

#define SCARD_ATTR_DEVICE_UNIT SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0001) /**< Instance of this vendor's reader attached to the computer. The first instance will be device unit 0, the next will be unit 1 (if it is the same brand of reader) and so on. Two different brands of readers will both have zero for this value. */
#define SCARD_ATTR_DEVICE_IN_USE SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0002) /**< Reserved for future use. */
#define SCARD_ATTR_DEVICE_FRIENDLY_NAME_A SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0003)
#define SCARD_ATTR_DEVICE_SYSTEM_NAME_A SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0004)
#define SCARD_ATTR_DEVICE_FRIENDLY_NAME_W SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0005)
#define SCARD_ATTR_DEVICE_SYSTEM_NAME_W SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0006)
#define SCARD_ATTR_SUPRESS_T1_IFS_REQUEST SCARD_ATTR_VALUE(SCARD_CLASS_SYSTEM, 0x0007) /**< FIXME */

#ifdef UNICODE
#define SCARD_ATTR_DEVICE_FRIENDLY_NAME SCARD_ATTR_DEVICE_FRIENDLY_NAME_W /**< Reader's display name. */
#define SCARD_ATTR_DEVICE_SYSTEM_NAME SCARD_ATTR_DEVICE_SYSTEM_NAME_W /**< Reader's system name. */
#else
#define SCARD_ATTR_DEVICE_FRIENDLY_NAME SCARD_ATTR_DEVICE_FRIENDLY_NAME_A /**< Reader's display name. */
#define SCARD_ATTR_DEVICE_SYSTEM_NAME SCARD_ATTR_DEVICE_SYSTEM_NAME_A /**< Reader's system name. */
#endif

/**
 * Provide source compatibility on different platforms
 */
#define SCARD_CTL_CODE(code) (0x42000000 + (code))

/**
 * PC/SC part 10 v2.02.07 March 2010 reader tags
 */
#define CM_IOCTL_GET_FEATURE_REQUEST SCARD_CTL_CODE(3400)

#define FEATURE_VERIFY_PIN_START         0x01
#define FEATURE_VERIFY_PIN_FINISH        0x02
#define FEATURE_MODIFY_PIN_START         0x03
#define FEATURE_MODIFY_PIN_FINISH        0x04
#define FEATURE_GET_KEY_PRESSED          0x05
#define FEATURE_VERIFY_PIN_DIRECT        0x06 /**< Verify PIN */
#define FEATURE_MODIFY_PIN_DIRECT        0x07 /**< Modify PIN */
#define FEATURE_MCT_READER_DIRECT        0x08
#define FEATURE_MCT_UNIVERSAL            0x09
#define FEATURE_IFD_PIN_PROPERTIES       0x0A /**< retrieve properties of the IFD regarding PIN handling */
#define FEATURE_ABORT                    0x0B
#define FEATURE_SET_SPE_MESSAGE          0x0C
#define FEATURE_VERIFY_PIN_DIRECT_APP_ID 0x0D
#define FEATURE_MODIFY_PIN_DIRECT_APP_ID 0x0E
#define FEATURE_WRITE_DISPLAY            0x0F
#define FEATURE_GET_KEY                  0x10
#define FEATURE_IFD_DISPLAY_PROPERTIES   0x11
#define FEATURE_GET_TLV_PROPERTIES       0x12
#define FEATURE_CCID_ESC_COMMAND         0x13
#define FEATURE_EXECUTE_PACE             0x20

/* structures used (but not defined) in PC/SC Part 10:
 * "IFDs with Secure Pin Entry Capabilities" */

#include <inttypes.h>

/* Set structure elements aligment on bytes
 * http://gcc.gnu.org/onlinedocs/gcc/Structure_002dPacking-Pragmas.html */
#if defined(__APPLE__) | defined(sun)
#pragma pack(1)
#else
#pragma pack(push, 1)
#endif

/** the structure must be 6-bytes long */
typedef struct
{
	uint8_t tag;
	uint8_t length;
	uint32_t value;	/**< This value is always in BIG ENDIAN format as documented in PCSC v2 part 10 ch 2.2 page 2. You can use ntohl() for example */
} PCSC_TLV_STRUCTURE;

/** Since CCID 1.4.1 (revision 5252) the byte order is no more important
 * These macros are now deprecated and should be removed in the future */
#define HOST_TO_CCID_16(x) (x)
#define HOST_TO_CCID_32(x) (x)

/** structure used with \ref FEATURE_VERIFY_PIN_DIRECT */
typedef struct
{
	uint8_t bTimerOut;	/**< timeout is seconds (00 means use default timeout) */
	uint8_t bTimerOut2; /**< timeout in seconds after first key stroke */
	uint8_t bmFormatString; /**< formatting options */
	uint8_t bmPINBlockString; /**< bits 7-4 bit size of PIN length in APDU,
	                        * bits 3-0 PIN block size in bytes after
	                        * justification and formatting */
	uint8_t bmPINLengthFormat; /**< bits 7-5 RFU,
	                         * bit 4 set if system units are bytes, clear if
	                         * system units are bits,
	                         * bits 3-0 PIN length position in system units */
	uint16_t wPINMaxExtraDigit; /**< 0xXXYY where XX is minimum PIN size in digits,
	                            and YY is maximum PIN size in digits */
	uint8_t bEntryValidationCondition; /**< Conditions under which PIN entry should
	                                 * be considered complete */
	uint8_t bNumberMessage; /**< Number of messages to display for PIN verification */
	uint16_t wLangId; /**< Language for messages */
	uint8_t bMsgIndex; /**< Message index (should be 00) */
	uint8_t bTeoPrologue[3]; /**< T=1 block prologue field to use (fill with 00) */
	uint32_t ulDataLength; /**< length of Data to be sent to the ICC */
	uint8_t abData
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L)
	[] /* valid C99 code */
#else
	[0] /* non-standard, but usually working code */
#endif
	; /**< Data to send to the ICC */
} PIN_VERIFY_STRUCTURE;

/** structure used with \ref FEATURE_MODIFY_PIN_DIRECT */
typedef struct
{
	uint8_t bTimerOut;	/**< timeout is seconds (00 means use default timeout) */
	uint8_t bTimerOut2; /**< timeout in seconds after first key stroke */
	uint8_t bmFormatString; /**< formatting options */
	uint8_t bmPINBlockString; /**< bits 7-4 bit size of PIN length in APDU,
	                        * bits 3-0 PIN block size in bytes after
	                        * justification and formatting */
	uint8_t bmPINLengthFormat; /**< bits 7-5 RFU,
	                         * bit 4 set if system units are bytes, clear if
	                         * system units are bits,
	                         * bits 3-0 PIN length position in system units */
	uint8_t bInsertionOffsetOld; /**< Insertion position offset in bytes for
	                             the current PIN */
	uint8_t bInsertionOffsetNew; /**< Insertion position offset in bytes for
	                             the new PIN */
	uint16_t wPINMaxExtraDigit;
	                         /**< 0xXXYY where XX is minimum PIN size in digits,
	                            and YY is maximum PIN size in digits */
	uint8_t bConfirmPIN; /**< Flags governing need for confirmation of new PIN */
	uint8_t bEntryValidationCondition; /**< Conditions under which PIN entry should
	                                 * be considered complete */
	uint8_t bNumberMessage; /**< Number of messages to display for PIN verification*/
	uint16_t wLangId; /**< Language for messages */
	uint8_t bMsgIndex1; /**< index of 1st prompting message */
	uint8_t bMsgIndex2; /**< index of 2d prompting message */
	uint8_t bMsgIndex3; /**< index of 3d prompting message */
	uint8_t bTeoPrologue[3]; /**< T=1 block prologue field to use (fill with 00) */
	uint32_t ulDataLength; /**< length of Data to be sent to the ICC */
	uint8_t abData
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L)
	[] /* valid C99 code */
#else
	[0] /* non-standard, but usually working code */
#endif
	; /**< Data to send to the ICC */
} PIN_MODIFY_STRUCTURE;

/** structure used with \ref FEATURE_IFD_PIN_PROPERTIES */
typedef struct {
	uint16_t wLcdLayout; /**< display characteristics */
	uint8_t bEntryValidationCondition;
	uint8_t bTimeOut2;
} PIN_PROPERTIES_STRUCTURE;

/* restore default structure elements alignment */
#if defined(__APPLE__) | defined(sun)
#pragma pack()
#else
#pragma pack(pop)
#endif

/* properties returned by FEATURE_GET_TLV_PROPERTIES */
#define PCSCv2_PART10_PROPERTY_wLcdLayout 1
#define PCSCv2_PART10_PROPERTY_bEntryValidationCondition 2
#define PCSCv2_PART10_PROPERTY_bTimeOut2 3
#define PCSCv2_PART10_PROPERTY_wLcdMaxCharacters 4
#define PCSCv2_PART10_PROPERTY_wLcdMaxLines 5
#define PCSCv2_PART10_PROPERTY_bMinPINSize 6
#define PCSCv2_PART10_PROPERTY_bMaxPINSize 7
#define PCSCv2_PART10_PROPERTY_sFirmwareID 8
#define PCSCv2_PART10_PROPERTY_bPPDUSupport 9
#define PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize 10
#define PCSCv2_PART10_PROPERTY_wIdVendor 11
#define PCSCv2_PART10_PROPERTY_wIdProduct 12

#endif

